src/core/store/common/tokensByAsset.selectors.ts

```diff
@@ -0,0 +1,86 @@
+import { createSelector } from '@reduxjs/toolkit';
+import { memoize } from 'lodash';
+
+import { toBN } from '@utils';
+import { getConfig } from '@config';
+import { TokenView, Address } from '@types';
+
+import { VaultsSelectors } from '../modules/vaults/vaults.selectors';
+import { LabsSelectors } from '../modules/labs/labs.selectors';
+import { TokensSelectors } from '../modules/tokens/tokens.selectors';
+import { AppSelectors } from '../modules/app/app.selectors';
+import { createToken } from '../modules/tokens/tokens.selectors';
+
+const { selectVaultsMap } = VaultsSelectors;
+const { selectLabsMap } = LabsSelectors;
+const { selectTokensMap, selectTokensUser } = TokensSelectors;
+const { selectServicesEnabled } = AppSelectors;
+
+export const selectDepositTokenOptionsByAsset = createSelector(
+  [selectVaultsMap, selectLabsMap, selectTokensMap, selectTokensUser, selectServicesEnabled],
+  (vaultsMap, labsMap, tokensMap, tokensUser, servicesEnabled) =>
+    memoize((assetAddress?: string): TokenView[] => {
+      if (!assetAddress) return [];
+
+      const { userTokensAddresses, userTokensMap, userTokensAllowancesMap } = tokensUser;
+      const assetData = vaultsMap[assetAddress] ? vaultsMap[assetAddress] : labsMap[assetAddress];
+      if (!assetData) return [];
+
+      const zapperDisabled = !servicesEnabled.zapper && assetData.metadata.zapInWith === 'zapperZapIn';
+      const depositTokenAddresses: Address[] = [];
+      let mainVaultToken: Address;
+      if (zapperDisabled) {
+        mainVaultToken = assetData.token;
+        depositTokenAddresses.push(mainVaultToken);
+      } else {
+        mainVaultToken = assetData.metadata.defaultDisplayToken;
+        depositTokenAddresses.push(mainVaultToken);
+        depositTokenAddresses.push(...userTokensAddresses.filter((address) => address !== mainVaultToken));
+      }
+
+      const tokens = depositTokenAddresses
+        .filter((address) => !!tokensMap[address])
+        .map((address) => {
+          const tokenData = tokensMap[address];
+          const userTokenData = userTokensMap[address];
+          const allowancesMap = userTokensAllowancesMap[address] ?? {};
+          return createToken({ tokenData, userTokenData, allowancesMap });
+        });
+      return tokens.filter((token) => toBN(token.balance).gt(0) || token.address === mainVaultToken);
+    })
+);
+
+export const selectWithdrawTokenOptionsByAsset = createSelector(
```

Comment:
Both of these functions seem to have quite a bit in common, could we extract some of the logic to a common function so that we can then refactor more easily?

<img width="2022" alt="Screen Shot 2022-05-18 at 9 54 58 am" src="https://user-images.githubusercontent.com/78794805/168930641-063e89d1-8175-44d6-8ca6-b672c4cadf0f.png">

Reply:
maybe the last part of both where `tokens` are mapped, but not sure its worth it.

Reply:
Let's leave it, usually there's a rule of 3... when it's repeated 3x it's the time to refactor.